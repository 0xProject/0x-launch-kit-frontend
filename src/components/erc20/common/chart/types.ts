export interface Cord {
    x: number;
    y: number;
}
export enum OrderSide {
    Bid,
    Ask,
}

export interface Margin {
    left: number;
    right: number;
    top: number;
    bottom: number;
}

export interface ChartDimensions {
    height: number;
    margin: Margin;
}
