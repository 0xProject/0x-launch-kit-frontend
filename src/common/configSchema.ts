import { Schema } from 'jsonschema';

export const configSchema: Schema = {
    id: '/configSchema',
    properties: {
        tokens: { type: 'array', items: { $ref: '/tokenMetaDataSchema' } },
        pairs: { type: 'array', items: { $ref: '/pairSchema' } },
    },
    required: ['tokens', 'pairs'],
    type: 'object',
};

export const schemas: Schema[] = [
    {
        id: '/addressSchema',
        type: 'string',
        pattern: '^0x[0-9a-f]{40}$',
    },
    {
        id: '/wholeNumberSchema',
        anyOf: [
            {
                type: 'string',
                pattern: '^\\d+$',
            },
            {
                type: 'integer',
            },
        ],
    },
    {
        id: '/marketFilterSchema',
        properties: {
            text: { type: 'string' },
            valute: { type: 'string' },
        },
        required: ['text', 'value'],
        type: 'object',
    },
    {
        id: '/pairSchema',
        properties: {
            base: { type: 'string' },
            quote: { type: 'string' },
        },
        required: ['base', 'quote'],
        type: 'object',
    },
    {
        id: '/tokenMetaDataSchema',
        properties: {
            symbol: { type: 'string' },
            name: { type: 'string' },
            icon: { type: 'string' },
            primaryColor: { type: 'string' },
            decimals: { $ref: '/wholeNumberSchema' },
            displayDecimals: { $ref: '/wholeNumberSchema' },
        },
        required: ['decimals', 'symbol', 'name', 'addresses'],
        type: 'object',
    },
    {
        id: '/configSchema',
        properties: {
            tokens: { type: 'array', items: { $ref: '/tokenMetaDataSchema' } },
            pairs: { type: 'array', items: { $ref: '/pairSchema' } },
            marketFilters: { type: 'array', items: { $ref: '/marketFilterSchema' } },
        },
        required: ['tokens', 'pairs'],
        type: 'object',
    },
];
