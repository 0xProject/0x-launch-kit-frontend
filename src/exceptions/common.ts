export const INSUFFICIENT_MAKER_BALANCE_ERR = 'INSUFFICIENT MAKER BALANCE';
export const SIGNATURE_ERR = 'User denied message signature';
export const USER_DENIED_TRANSACTION_SIGNATURE_ERR = 'User denied transaction signature';
export const INSUFFICIENT_ORDERS_TO_FILL_AMOUNT_ERR = 'There are not enough orders to fill this amount';
export const RELAYER_ERR = 'There was an error with the relayer';
export const INSUFFICIENT_FEE_BALANCE = 'INSUFFICIENT MAKER FEE BALANCE';
export const INSUFFICIENT_FEE_BALANCE_MSG = `You don't have enough 0x to pay fees`;
export const INSUFFICIENT_ETH_BALANCE_FOR_DEPOSIT = 'INSUFFICIENT_ETH_BALANCE_FOR_DEPOSIT';
export const UNEXPECTED_ERROR = 'An unexpected error happened';
