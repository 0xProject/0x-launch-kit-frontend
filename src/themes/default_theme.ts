import { Theme, ThemeModalStyle, ThemeProperties } from './commons';

const modalThemeStyle: ThemeModalStyle = {
    content: {
        backgroundColor: '#fff',
        borderColor: '#dedede',
        bottom: 'auto',
        display: 'flex',
        flexDirection: 'column',
        flexGrow: '0',
        left: 'auto',
        maxHeight: '90%',
        minWidth: '350px',
        overflow: 'hidden',
        padding: '16px',
        position: 'relative',
        right: 'auto',
        top: 'auto',
    },
    overlay: {
        alignItems: 'center',
        backgroundColor: 'rgba(0, 0, 0, 0.4)',
        display: 'flex',
        justifyContent: 'center',
        zIndex: '12345',
    },
};

const lightThemeColors: ThemeProperties = {
    background: '#f4f3f4',
    backgroundERC721: '#F9FAFC',
    borderColor: '#dedede',
    boxShadow: '0 10px 10px rgba(0, 0, 0, 0.1)',
    buttonBuyBackgroundColor: '#3CB34F',
    buttonCollectibleSellBackgroundColor: '#00AE99',
    buttonConvertBackgroundColor: '#fff',
    buttonConvertBorderColor: '#dedede',
    buttonConvertTextColor: '#474747',
    buttonErrorBackgroundColor: '#FF6534',
    buttonPrimaryBackgroundColor: '#002979',
    buttonQuaternaryBackgroundColor: '#00AE99',
    buttonSecondaryBackgroundColor: '#474747',
    buttonSellBackgroundColor: '#FF6534',
    buttonTertiaryBackgroundColor: '#F6851B',
    buttonTextColor: '#fff',
    cardImageBackgroundColor: '#EBF0F5',
    cardBackgroundColor: '#fff',
    cardBorderColor: '#dedede',
    cardTitleColor: '#000',
    cardTitleOwnerColor: '#3CB34F',
    chartColor: '#00AE99',
    darkBlue: '#002979',
    darkGray: '#474747',
    darkerGray: '#666',
    dropdownBackgroundColor: '#fff',
    dropdownBorderColor: '#dedede',
    dropdownTextColor: '#000',
    errorButtonBackground: '#FF6534',
    errorCardBackground: '#FAF4EF',
    errorCardBorder: '#F39E4B',
    errorCardText: '#F68C24',
    ethBoxActiveColor: '#002979',
    ethBoxBorderColor: '#dedede',
    ethSetMinEthButtonBorderColor: '#000',
    ethSliderThumbBorderColor: 'rgba(0, 0, 0, 0.142)',
    ethSliderThumbColor: '#fff',
    gray: '#808080',
    green: '#3CB34F',
    iconLockedColor: '#000',
    iconUnlockedColor: '#C4C4C4',
    inactiveTabBackgroundColor: '#f9f9f9',
    lightGray: '#B9B9B9',
    logoERC20Color: '#0029FF',
    logoERC20TextColor: '#000',
    logoERC721Color: '#00AE99',
    logoERC721TextColor: '#000',
    marketsSearchFieldBackgroundColor: '#eaeaea',
    marketsSearchFieldBorderColor: '#dedede',
    marketsSearchFieldTextColor: '#333',
    modalSearchFieldBackgroundColor: '#fff',
    modalSearchFieldBorderColor: '#fff',
    modalSearchFieldPlaceholderColor: '#DEDEDE',
    modalSearchFieldTextColor: '#000',
    myWalletLinkColor: '#333333',
    notificationActive: '#F8F8F8',
    notificationIconColor: '#474747',
    notificationsBadgeColor: '#ff6534',
    numberDecimalsColor: '#dedede',
    red: '#FF6534',
    rowActive: '#FBFDFF',
    rowOrderActive: '#dedede',
    simplifiedTextBoxColor: '#F9F9F9',
    stepsProgressCheckMarkColor: '#fff',
    stepsProgressStartingDotColor: '#000',
    stepsProgressStepLineColor: 'rgba(0, 0, 0, 0.1)',
    stepsProgressStepLineProgressColor: '#000',
    stepsProgressStepTitleColor: '#e6e6e6',
    stepsProgressStepTitleColorActive: '#000',
    tableBorderColor: '#dedede',
    tdColor: '#000',
    textColorCommon: '#000',
    textDark: '#666',
    textInputBackgroundColor: '#F9F9F9',
    textInputBorderColor: '#dedede',
    textInputTextColor: '#000',
    textLight: '#999',
    textLighter: '#666',
    thColor: '#B9B9B9',
    tooltipBackgroundColor: '#222',
    tooltipTextColor: '#fff',
    topbarBackgroundColor: '#fff',
    topbarBorderColor: '#dedede',
    topbarSeparatorColor: '#dedede',
};

export class DefaultTheme implements Theme {
    public componentsTheme: ThemeProperties;
    public modalTheme: ThemeModalStyle;
    constructor() {
        this.componentsTheme = lightThemeColors;
        this.modalTheme = modalThemeStyle;
    }
}
