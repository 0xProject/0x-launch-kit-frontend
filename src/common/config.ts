import { Validator } from 'jsonschema';

import configFile from '../config.json';
import { ConfigFile } from '../util/types';

import { configSchema, schemas } from './configSchema';

export class Config {
    private static _instance: Config;
    private readonly _validator: Validator;
    private readonly _config: ConfigFile;
    public static getInstance(): Config {
        if (!Config._instance) {
            Config._instance = new Config();
        }
        return Config._instance;
    }
    public static getConfig(): ConfigFile {
        return this.getInstance()._config;
    }
    constructor() {
        this._validator = new Validator();
        for (const schema of schemas) {
            this._validator.addSchema(schema, schema.id);
        }
        this._validator.validate(configFile, configSchema, { throwError: true });
        this._config = configFile;
    }
}
