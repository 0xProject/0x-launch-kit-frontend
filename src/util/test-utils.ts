import { assetDataUtils, BigNumber, OrderStatus } from '0x.js';
import { SignedOrder } from '@0x/connect';
import * as Factory from 'factory.ts';

import { TokenMetaData } from '../common/tokens_meta_data';

import { Collectible, OrderSide, Token, TokenBalance, UIOrder } from './types';

export const makeOrder = ({
    makerAssetAmount,
    takerAssetAmount,
    makerTokenAddress,
    takerTokenAddress,
}: {
    makerAssetAmount: string;
    takerAssetAmount: string;
    makerTokenAddress: string;
    takerTokenAddress: string;
}): SignedOrder => {
    const oneYearFromNow = Math.floor(new Date().valueOf() / 1000) + 3600 * 24 * 365;

    return {
        makerAddress: '0x5409ed021d9299bf6814279a6a1411a7e866a631',
        exchangeAddress: '0x48bacb9266a570d521063ef5dd96e61686dbe788',
        takerAddress: '0x0000000000000000000000000000000000000000',
        senderAddress: '0x0000000000000000000000000000000000000000',
        feeRecipientAddress: '0x0000000000000000000000000000000000000000',
        expirationTimeSeconds: new BigNumber(oneYearFromNow),
        salt: new BigNumber('4253024946184612711043400606072358480456814389671017608724053124375087745'),
        makerAssetAmount: new BigNumber(makerAssetAmount),
        takerAssetAmount: new BigNumber(takerAssetAmount),
        makerAssetData: assetDataUtils.encodeERC20AssetData(makerTokenAddress),
        takerAssetData: assetDataUtils.encodeERC20AssetData(takerTokenAddress),
        makerFee: new BigNumber('0'),
        takerFee: new BigNumber('0'),
        signature: '',
    };
};

export const uiOrder = (params = {}): UIOrder => {
    const rawOrder: any = {};
    return {
        filled: new BigNumber(0),
        price: new BigNumber(1),
        rawOrder,
        side: OrderSide.Sell,
        size: new BigNumber(1),
        status: OrderStatus.Fillable,
        ...params,
    };
};

export const openOrder = (params = {}): UIOrder => {
    return uiOrder({
        status: OrderStatus.Fillable,
        ...params,
    });
};

export const closedOrder = (params = {}): UIOrder => {
    return uiOrder({
        status: OrderStatus.FullyFilled,
        ...params,
    });
};

const repeat = (x: any, times: number) => [...Array(times)].map(() => x);

export const addressFactory = Factory.Sync.makeFactory<{ address: string }>({
    address: Factory.each((i: number) => {
        const prefix = '0x';
        const zeros = repeat('0', 39 - Math.floor(Math.log(i + 1) / Math.log(10))).join('');
        const suffix = (i + 1).toString();

        return `${prefix}${zeros}${suffix}`;
    }),
});

export const tokenFactory = Factory.Sync.makeFactory<Token>({
    address: Factory.each(() => addressFactory.build().address),
    decimals: 0,
    displayDecimals: 2,
    name: Factory.each(i => `Mock Token ${i}`),
    primaryColor: '#ff0000',
    // @ts-ignore
    symbol: Factory.each(i => `MOCK${i}`),
    icon: undefined,
});

export const tokenMetaDataFactory = Factory.Sync.makeFactory<TokenMetaData>({
    addresses: Factory.each(() => ({
        50: addressFactory.build().address,
    })),
    decimals: 0,
    name: Factory.each(i => `Mock Token ${i}`),
    primaryColor: '#ff0000',
    // @ts-ignore
    symbol: Factory.each(i => `MOCK${i}`),
    icon: '',
});

export const tokenBalanceFactory = Factory.Sync.makeFactory<TokenBalance>({
    balance: new BigNumber(2),
    isUnlocked: true,
    token: tokenFactory.build(),
});

export const collectibleFactory = Factory.Sync.makeFactory<Collectible>({
    assetUrl: '',
    color: '',
    currentOwner: addressFactory.build().address,
    description: Factory.each(i => `Mocked collectible ${i}`),
    image: '',
    name: Factory.each(i => `NFT ${i}`),
    order: null,
    tokenId: Factory.each(i => i.toString()),
});
