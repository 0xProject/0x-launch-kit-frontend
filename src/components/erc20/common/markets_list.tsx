import React, { HTMLAttributes } from 'react';
import { connect } from 'react-redux';
import styled from 'styled-components';

import { UI_DECIMALS_DISPLAYED_SPREAD_PERCENT, USE_RELAYER_MARKET_UPDATES } from '../../../common/constants';
import { marketFilters } from '../../../common/markets';
import { changeMarket, goToHome } from '../../../store/actions';
import { getBaseToken, getCurrencyPair, getMarkets, getQuoteToken, getWeb3State } from '../../../store/selectors';
import { themeBreakPoints, themeDimensions } from '../../../themes/commons';
import { getKnownTokens } from '../../../util/known_tokens';
import { filterMarketsByString, filterMarketsByTokenSymbol } from '../../../util/markets';
import { formatTokenSymbol } from '../../../util/tokens';
import { CurrencyPair, Filter, Market, StoreState, Token, Web3State } from '../../../util/types';
import { Card } from '../../common/card';
import { EmptyContent } from '../../common/empty_content';
import { MagnifierIcon } from '../../common/icons/magnifier_icon';
import { TokenIcon } from '../../common/icons/token_icon';
import { LoadingWrapper } from '../../common/loading';
import { CustomTDFirst, CustomTDLast, Table, TBody, THead, THFirst, THLast, TR } from '../../common/table';

interface PropsDivElement extends HTMLAttributes<HTMLDivElement> {}

interface DispatchProps {
    changeMarket: (currencyPair: CurrencyPair) => any;
    goToHome: () => any;
}

interface PropsToken {
    baseToken: Token | null;
    currencyPair: CurrencyPair;
    markets: Market[] | null;
    web3State?: Web3State;
    quoteToken: Token | null;
}

type Props = PropsDivElement & PropsToken & DispatchProps;

interface State {
    selectedFilter: Filter;
    search: string;
}

interface TokenFiltersTabProps {
    active: boolean;
    onClick: number;
}

interface MarketRowProps {
    active: boolean;
}

const rowHeight = '48px';

const MarketListCard = styled(Card)`
    height: 100%;
    overflow: auto;
    margin-top: 5px;
`;

const MarketsFilters = styled.div`
    align-items: center;
    border-bottom: 1px solid ${props => props.theme.componentsTheme.dropdownBorderColor};
    display: flex;
    justify-content: space-between;
    min-height: ${rowHeight};
    padding: 8px 8px 8px ${themeDimensions.horizontalPadding};
`;

const TokenFiltersTabs = styled.div`
    align-items: center;
    display: flex;
    margin-right: 10px;
`;

const TokenFiltersTab = styled.span<TokenFiltersTabProps>`
    color: ${props =>
        props.active ? props.theme.componentsTheme.textColorCommon : props.theme.componentsTheme.lightGray};
    cursor: pointer;
    font-size: 14px;
    font-weight: 500;
    line-height: 1.2;
    user-select: none;

    &:after {
        color: ${props => props.theme.componentsTheme.lightGray};
        content: '/';
        margin: 0 6px;
    }

    &:last-child:after {
        display: none;
    }
`;

const searchFieldHeight = '32px';
const searchFieldWidth = '142px';

const SearchWrapper = styled.div`
    height: ${searchFieldHeight};
    position: relative;
    width: ${searchFieldWidth};
`;

const SearchField = styled.input`
    background: ${props => props.theme.componentsTheme.marketsSearchFieldBackgroundColor};
    border-radius: ${themeDimensions.borderRadius};
    border: 1px solid ${props => props.theme.componentsTheme.marketsSearchFieldBorderColor};
    color: ${props => props.theme.componentsTheme.marketsSearchFieldTextColor};
    font-size: 13px;
    height: ${searchFieldHeight};
    left: 0;
    outline: none;
    padding: 0 15px 0 30px;
    position: absolute;
    top: 0;
    width: ${searchFieldWidth};
    z-index: 1;

    &:focus {
        border-color: ${props => props.theme.componentsTheme.marketsSearchFieldBorderColor};
    }
`;

const MagnifierIconWrapper = styled.div`
    line-height: 30px;
    height: 100%;
    left: 11px;
    position: absolute;
    top: 0;
    width: 14px;
    z-index: 12;
`;

const TableWrapper = styled.div`
    max-height: 520px;
    overflow: auto;
    position: relative;
`;

const verticalCellPadding = `
    padding-bottom: 10px;
    padding-top: 10px;
`;

const tableHeaderFontWeight = `
    font-weight: 700;
`;

const TRStyled = styled(TR)<MarketRowProps>`
    background-color: ${props => (props.active ? props.theme.componentsTheme.rowActive : 'transparent')};
    cursor: ${props => (props.active ? 'default' : 'pointer')};

    &:hover {
        background-color: ${props => props.theme.componentsTheme.rowActive};
    }

    &:last-child > td {
        border-bottom-left-radius: ${themeDimensions.borderRadius};
        border-bottom-right-radius: ${themeDimensions.borderRadius};
        border-bottom: none;
    }
`;

// Has a special left-padding: needs a specific selector to override the theme
const THFirstStyled = styled(THFirst)`
    ${verticalCellPadding}
    ${tableHeaderFontWeight}

    &, &:last-child {
        padding-left: 21.6px;
    }
`;

const THLastStyled = styled(THLast)`
    ${verticalCellPadding};
    ${tableHeaderFontWeight}
`;

const CustomTDFirstStyled = styled(CustomTDFirst)`
    ${verticalCellPadding};
`;

const CustomTDLastStyled = styled(CustomTDLast)`
    ${verticalCellPadding};
`;

const TokenIconAndLabel = styled.div`
    align-items: center;
    display: flex;
    justify-content: flex-start;
    @media (max-width: ${themeBreakPoints.md}) {
        flex-wrap: wrap;
        justify-content: center;
    }
`;

const TokenLabel = styled.div`
    color: ${props => props.theme.componentsTheme.textColorCommon};
    font-size: 14px;
    font-weight: 700;
    line-height: 1.2;
    margin: 0 0 0 12px;
`;

class MarketsList extends React.Component<Props, State> {
    public readonly state: State = {
        selectedFilter: marketFilters[0],
        search: '',
    };

    public render = () => {
        const { baseToken, quoteToken, web3State } = this.props;
        let content: React.ReactNode;
        const defaultBehaviour = () => {
            if (web3State !== Web3State.Error && (!baseToken || !quoteToken)) {
                content = <LoadingWrapper minHeight="120px" />;
            } else if (!baseToken || !quoteToken) {
                content = <EmptyContent alignAbsoluteCenter={true} text="There are no market details to show" />;
            } else {
                content = (
                    <>
                        <MarketsFilters>
                            {/*  <MarketsFiltersLabel>Markets</MarketsFiltersLabel>*/}
                            {this._getTokensFilterTabs()}
                            {this._getSearchField()}
                        </MarketsFilters>
                        <TableWrapper>{this._getMarkets()}</TableWrapper>
                    </>
                );
            }
        };
        if (USE_RELAYER_MARKET_UPDATES) {
            defaultBehaviour();
        } else {
            switch (web3State) {
                case Web3State.Locked:
                case Web3State.NotInstalled: {
                    content = <EmptyContent alignAbsoluteCenter={true} text="There are no market details to show" />;
                    break;
                }
                case Web3State.Loading: {
                    content = <LoadingWrapper minHeight="120px" />;
                    break;
                }
                default: {
                    defaultBehaviour();
                    break;
                }
            }
        }

        return <MarketListCard title="Markets">{content}</MarketListCard>;
    };

    private readonly _getTokensFilterTabs = () => {
        return (
            <TokenFiltersTabs>
                {marketFilters.map((filter: Filter, index) => {
                    return (
                        <TokenFiltersTab
                            active={filter === this.state.selectedFilter}
                            key={index}
                            onClick={this._setTokensFilterTab.bind(this, filter)}
                        >
                            {filter.text}
                        </TokenFiltersTab>
                    );
                })}
            </TokenFiltersTabs>
        );
    };

    private readonly _setTokensFilterTab: any = (filter: Filter) => {
        this.setState({ selectedFilter: filter });
    };

    private readonly _getSearchField = () => {
        return (
            <SearchWrapper>
                <MagnifierIconWrapper>{MagnifierIcon()}</MagnifierIconWrapper>
                <SearchField onChange={this._handleChange} value={this.state.search} />
            </SearchWrapper>
        );
    };

    private readonly _handleChange = (e: any) => {
        const search = e.currentTarget.value;

        this.setState({
            search,
        });
    };

    private readonly _getMarkets = () => {
        const { baseToken, currencyPair, markets } = this.props;
        const { search, selectedFilter } = this.state;

        if (!baseToken || !markets) {
            return null;
        }

        const filteredMarkets =
            selectedFilter == null || selectedFilter.value === null
                ? markets
                : filterMarketsByTokenSymbol(markets, selectedFilter.value);
        const searchedMarkets = filterMarketsByString(filteredMarkets, search);

        return (
            <Table>
                <THead>
                    <TR>
                        <THFirstStyled styles={{ textAlign: 'left' }}>Market</THFirstStyled>
                        <THLastStyled styles={{ textAlign: 'center' }}>Best Ask</THLastStyled>
                        <THLastStyled styles={{ textAlign: 'center' }}>Best Bid</THLastStyled>
                        <THLastStyled styles={{ textAlign: 'center' }}>Spread</THLastStyled>
                    </TR>
                </THead>
                <TBody>
                    {searchedMarkets.map((market, index) => {
                        const isActive =
                            market.currencyPair.base === currencyPair.base &&
                            market.currencyPair.quote === currencyPair.quote;
                        const setSelectedMarket = () => this._setSelectedMarket(market.currencyPair);

                        const token = getKnownTokens().getTokenBySymbol(market.currencyPair.base);

                        const baseSymbol = formatTokenSymbol(market.currencyPair.base).toUpperCase();
                        const quoteSymbol = formatTokenSymbol(market.currencyPair.quote).toUpperCase();

                        return (
                            <TRStyled active={isActive} key={index} onClick={setSelectedMarket}>
                                <CustomTDFirstStyled styles={{ textAlign: 'left', borderBottom: true }}>
                                    <TokenIconAndLabel>
                                        <TokenIcon
                                            symbol={token.symbol}
                                            primaryColor={token.primaryColor}
                                            icon={token.icon}
                                        />
                                        <TokenLabel>
                                            {baseSymbol} / {quoteSymbol}
                                        </TokenLabel>
                                    </TokenIconAndLabel>
                                </CustomTDFirstStyled>
                                <CustomTDLastStyled styles={{ textAlign: 'center', borderBottom: true, tabular: true }}>
                                    {this._getBestAsk(market)}
                                </CustomTDLastStyled>
                                <CustomTDLastStyled styles={{ textAlign: 'center', borderBottom: true, tabular: true }}>
                                    {this._getBestBid(market)}
                                </CustomTDLastStyled>
                                <CustomTDLastStyled styles={{ textAlign: 'center', borderBottom: true, tabular: true }}>
                                    {this._getSpreadInPercentage(market)}
                                </CustomTDLastStyled>
                            </TRStyled>
                        );
                    })}
                </TBody>
            </Table>
        );
    };

    private readonly _setSelectedMarket: any = (currencyPair: CurrencyPair) => {
        this.props.changeMarket(currencyPair);
        this.props.goToHome();
    };

    private readonly _getBestAsk: any = (market: Market) => {
        if (market.bestAsk) {
            return market.bestAsk.toFixed(market.currencyPair.config.pricePrecision);
        }

        return '-';
    };
    private readonly _getBestBid: any = (market: Market) => {
        if (market.bestBid) {
            return market.bestBid.toFixed(market.currencyPair.config.pricePrecision);
        }

        return '-';
    };
    private readonly _getSpreadInPercentage: any = (market: Market) => {
        if (market.spreadInPercentage) {
            return `${market.spreadInPercentage.toFixed(UI_DECIMALS_DISPLAYED_SPREAD_PERCENT)} %`;
        }

        return '-';
    };
}

const mapStateToProps = (state: StoreState): PropsToken => {
    return {
        baseToken: getBaseToken(state),
        currencyPair: getCurrencyPair(state),
        markets: getMarkets(state),
        quoteToken: getQuoteToken(state),
        web3State: getWeb3State(state),
    };
};

const mapDispatchToProps = (dispatch: any): DispatchProps => {
    return {
        changeMarket: (currencyPair: CurrencyPair) => dispatch(changeMarket(currencyPair)),
        goToHome: () => dispatch(goToHome()),
    };
};

const MarketsListContainer = connect(mapStateToProps, mapDispatchToProps)(MarketsList);

export { MarketsList, MarketsListContainer };
