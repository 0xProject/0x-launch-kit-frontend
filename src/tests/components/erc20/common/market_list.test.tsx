/**
 * @jest-environment jsdom
 */

import { BigNumber } from '0x.js';
import React from 'react';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import configureMockStore from 'redux-mock-store';
import { ThemeProvider } from 'styled-components';

import { MarketsList } from '../../../../components/erc20/common/markets_list';
import { DefaultTheme } from '../../../../themes/default_theme';
import { getCurrencyPairFromTokens, tokenFactory } from '../../../../util/test-utils';
import { Web3State } from '../../../../util/types';

const mockStore = configureMockStore([]);

const theme = new DefaultTheme();

const config = {
    basePrecision: 8,
    pricePrecision: 8,
    quotePrecision: 8,
    minAmount: 0,
    maxAmount: 1000000,
};

const marketData = {
    spreadInPercentage: new BigNumber(1),
    bestAsk: new BigNumber(1),
    bestBid: new BigNumber(1),
};

const marketExamples = {
    zrxWeth: {
        currencyPair: {
            base: 'zrx',
            quote: 'weth',
            config,
        },
        price: null,
        ...marketData,
    },
    wethZrx: {
        currencyPair: {
            base: 'weth',
            quote: 'zrx',
            config,
        },
        price: null,
        ...marketData,
    },
    daiMkr: {
        currencyPair: {
            base: 'dai',
            quote: 'mkr',
            config,
        },
        price: null,
        ...marketData,
    },
    daiWeth: {
        currencyPair: {
            base: 'dai',
            quote: 'weth',
            config,
        },
        price: null,
        ...marketData,
    },
};
const { zrxWeth, wethZrx, daiWeth } = marketExamples;

describe('MarketsList', () => {
    it('Renders market list', () => {
        const markets = [zrxWeth, wethZrx, daiWeth];

        const baseToken = tokenFactory.build();
        const quoteToken = tokenFactory.build();
        const currencyPair = getCurrencyPairFromTokens(baseToken, quoteToken);
        const changeMarket = jest.fn();
        const goToHome = jest.fn();

        const store = mockStore({});

        // when
        const wrapper = (
            <ThemeProvider theme={theme}>
                <Provider store={store}>
                    <MarketsList
                        baseToken={baseToken}
                        quoteToken={quoteToken}
                        currencyPair={currencyPair}
                        changeMarket={changeMarket}
                        goToHome={goToHome}
                        markets={markets}
                    />
                </Provider>
            </ThemeProvider>
        );

        const tree = renderer.create(wrapper).toJSON();

        // then
        expect(tree).toMatchSnapshot();
    });

    it('Should not render markets list if the user does not have metamask', () => {
        const markets = [zrxWeth, wethZrx, daiWeth];

        const baseToken = tokenFactory.build();
        const quoteToken = tokenFactory.build();
        const currencyPair = getCurrencyPairFromTokens(baseToken, quoteToken);
        const changeMarket = jest.fn();
        const goToHome = jest.fn();

        const store = mockStore({});

        // when
        const wrapper = (
            <ThemeProvider theme={theme}>
                <Provider store={store}>
                    <MarketsList
                        baseToken={baseToken}
                        quoteToken={quoteToken}
                        currencyPair={currencyPair}
                        changeMarket={changeMarket}
                        web3State={Web3State.NotInstalled}
                        goToHome={goToHome}
                        markets={markets}
                    />
                </Provider>
            </ThemeProvider>
        );

        const tree = renderer.create(wrapper).toJSON();

        // then
        expect(tree).toMatchSnapshot();
    });

    it('Should not render markets list if the user rejected the metamask permissions', () => {
        const markets = [zrxWeth, wethZrx, daiWeth];

        const baseToken = tokenFactory.build();
        const quoteToken = tokenFactory.build();
        const currencyPair = getCurrencyPairFromTokens(baseToken, quoteToken);
        const changeMarket = jest.fn();
        const goToHome = jest.fn();

        const store = mockStore({});

        // when
        const wrapper = (
            <ThemeProvider theme={theme}>
                <Provider store={store}>
                    <MarketsList
                        baseToken={baseToken}
                        quoteToken={quoteToken}
                        currencyPair={currencyPair}
                        changeMarket={changeMarket}
                        web3State={Web3State.Locked}
                        goToHome={goToHome}
                        markets={markets}
                    />
                </Provider>
            </ThemeProvider>
        );

        const tree = renderer.create(wrapper).toJSON();

        // then
        expect(tree).toMatchSnapshot();
    });
});
