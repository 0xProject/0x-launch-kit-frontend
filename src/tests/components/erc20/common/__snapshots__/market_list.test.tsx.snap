// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MarketsList Renders market list 1`] = `
<div
  className="sc-kpOJdX dBWysO sc-bwzfXH cfOWoU sc-bdVaJa fJeBhq"
>
  <div
    className="sc-htpNat fMPtTX"
  >
    <h1
      className="sc-bxivhb DvYdZ"
    >
      Markets
    </h1>
  </div>
  <div
    className="sc-ifAKCX jlvJWy"
  >
    <div
      className="sc-dxgOiQ hJUIXU"
    >
      <div
        className="sc-ckVGcZ gHifNV"
      >
        <span
          className="sc-jKJlTe fLRBMv"
          onClick={[Function]}
        >
          VSF
        </span>
        <span
          className="sc-jKJlTe hvcGSa"
          onClick={[Function]}
        >
          ETH
        </span>
        <span
          className="sc-jKJlTe hvcGSa"
          onClick={[Function]}
        >
          ALL
        </span>
      </div>
      <div
        className="sc-eNQAEJ hIMUcL"
      >
        <div
          className="sc-kEYyzF aVeFp"
        >
          <svg
            fill="none"
            height="13"
            viewBox="0 0 14 13"
            width="14"
            xmlns="http://www.w3.org/2000/svg"
          >
            <circle
              cx="5"
              cy="5"
              r="4"
              stroke="#BFBFBF"
              strokeLinejoin="round"
              strokeWidth="2"
            />
            <path
              d="M12.5 11.5L8 7"
              stroke="#BFBFBF"
              strokeWidth="2"
            />
          </svg>
        </div>
        <input
          className="sc-hMqMXs dXQhoL"
          onChange={[Function]}
          value=""
        />
      </div>
    </div>
    <div
      className="sc-kkGfuU gzuEEL"
    >
      <table
        className="sc-gZMcBi eiSZzD"
      >
        <thead
          className="sc-gqjmRU kjTfhg"
        >
          <tr
            className="sc-jTzLTM eCMmlJ"
          >
            <th
              className="sc-fjdhpX sc-kgoBCf sc-hSdWYo eplmLF"
            >
              Market
            </th>
            <th
              className="sc-fjdhpX sc-kGXeez sc-eHgmQL JqnTQ"
            >
              Best Ask
            </th>
            <th
              className="sc-fjdhpX sc-kGXeez sc-eHgmQL JqnTQ"
            >
              Best Bid
            </th>
            <th
              className="sc-fjdhpX sc-kGXeez sc-eHgmQL JqnTQ"
            >
              Spread
            </th>
          </tr>
        </thead>
        <tbody
          className="sc-VigVT fIhGog"
        />
      </table>
    </div>
  </div>
</div>
`;

exports[`MarketsList Should not render markets list if the user does not have metamask 1`] = `
<div
  className="sc-kpOJdX dBWysO sc-bwzfXH cfOWoU sc-bdVaJa fJeBhq"
>
  <div
    className="sc-htpNat fMPtTX"
  >
    <h1
      className="sc-bxivhb DvYdZ"
    >
      Markets
    </h1>
  </div>
  <div
    className="sc-ifAKCX jlvJWy"
  >
    <div
      className="sc-EHOje kXSMmk"
    >
      There are no market details to show
    </div>
  </div>
</div>
`;

exports[`MarketsList Should not render markets list if the user rejected the metamask permissions 1`] = `
<div
  className="sc-kpOJdX dBWysO sc-bwzfXH cfOWoU sc-bdVaJa fJeBhq"
>
  <div
    className="sc-htpNat fMPtTX"
  >
    <h1
      className="sc-bxivhb DvYdZ"
    >
      Markets
    </h1>
  </div>
  <div
    className="sc-ifAKCX jlvJWy"
  >
    <div
      className="sc-EHOje kXSMmk"
    >
      There are no market details to show
    </div>
  </div>
</div>
`;
